flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSource.java

```diff
@@ -0,0 +1,209 @@
+    public SplitEnumerator<HybridSourceSplit, HybridSourceEnumState> createEnumerator(
+            SplitEnumeratorContext<HybridSourceSplit> enumContext) {
+        return new HybridSourceSplitEnumerator(enumContext, sourceChain);
+    }
+
+    @Override
+    public SplitEnumerator<HybridSourceSplit, HybridSourceEnumState> restoreEnumerator(
+            SplitEnumeratorContext<HybridSourceSplit> enumContext, HybridSourceEnumState checkpoint)
+            throws Exception {
+        return new HybridSourceSplitEnumerator(
+                enumContext, sourceChain, checkpoint.getCurrentSourceIndex());
+    }
+
+    @Override
+    public SimpleVersionedSerializer<HybridSourceSplit> getSplitSerializer() {
+        List<SimpleVersionedSerializer<SourceSplit>> serializers = new ArrayList<>();
+        sourceChain.sources.forEach(
+                t -> serializers.add(castSerializer(t.f0.getSplitSerializer())));
+        return new SplitSerializerWrapper<>(serializers);
+    }
+
+    @Override
+    public SimpleVersionedSerializer<HybridSourceEnumState> getEnumeratorCheckpointSerializer() {
+        List<SimpleVersionedSerializer<Object>> serializers = new ArrayList<>();
+        sourceChain.sources.forEach(
+                t -> serializers.add(castSerializer(t.f0.getEnumeratorCheckpointSerializer())));
+        return new HybridSourceEnumStateSerializer(serializers);
+    }
+
+    private static <T> SimpleVersionedSerializer<T> castSerializer(
+            SimpleVersionedSerializer<? extends T> s) {
+        @SuppressWarnings("rawtypes")
+        SimpleVersionedSerializer s1 = s;
+        return s1;
+    }
+
+    /** Serializes splits by delegating to the source-indexed split serializer. */
+    public static class SplitSerializerWrapper<SplitT extends SourceSplit>
+            implements SimpleVersionedSerializer<HybridSourceSplit> {
+
+        final List<SimpleVersionedSerializer<SourceSplit>> serializers;
+
+        public SplitSerializerWrapper(List<SimpleVersionedSerializer<SourceSplit>> serializers) {
+            this.serializers = serializers;
+        }
+
+        @Override
+        public int getVersion() {
+            return 0;
+        }
+
+        @Override
+        public byte[] serialize(HybridSourceSplit split) throws IOException {
+            try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                    DataOutputStream out = new DataOutputStream(baos)) {
+                out.writeInt(split.sourceIndex());
+                out.writeInt(serializerOf(split.sourceIndex()).getVersion());
+                byte[] serializedSplit =
+                        serializerOf(split.sourceIndex()).serialize(split.getWrappedSplit());
+                out.writeInt(serializedSplit.length);
+                out.write(serializedSplit);
+                return baos.toByteArray();
+            }
+        }
+
+        @Override
+        public HybridSourceSplit deserialize(int version, byte[] serialized) throws IOException {
```

Comment:
Should the `deserialize ` method check the version？

Reply:
we can also move the remaining code to a `deserializeV0` method

Reply:
Done!